### PR TITLE
Update io_bazel_rules_go, bazel_gazelle, and rules_proto

### PR DESCRIPTION
**Pull request by SaveTheRbtz**
Updates in the diff:
- `io_bazel_rules_go` v0.27.0 -> v0.29.0
- `bazel_gazelle` v0.23.0 -> v0.24.0
- `rules_proto` 2019 -> 4.0.0

This diff moves `go_repository` rules out of `WORKSPACE` to a gazelle-controlled `deps.bzl`.

While here, added a `gazelle-update-repos` target for the ease of dependency update:
```
go get -u ./...
go mod tidy
bazel run //:gazelle-update-repos
bazel run //:gazelle
```

This upgrade enables using `rules_go`'s [IDE Integration](https://github.com/bazelbuild/rules_go/wiki/Editor-setup) through `@io_bazel_rules_go//go/tools/gopackagesdriver`.

### CODE REVIEW

**Comment by mostynb** [comment]
Hi, thanks for the cleanup. Normally I just update the modules the regular go way and then run `bazel run //:gazelle -- update-repos -from_file=go.mod` - what's the benefit of moving these to a separate file?

**Comment by SaveTheRbtz** [comment]
The main benefit is splitting the human-curated code (`WORKSPACE`) from the computer-generated one (`deps.bzl`).  With out this workspace file is quite hard to navigate: for example even right now docker rules are buried in the middle of the `go_repository` rules: 
https://github.com/buchgr/bazel-remote/blob/0bc183e484eeb11f403ee1ecea5cf7dd759144bb/WORKSPACE#L195-L227

PS. Most monorepos end up with this model over time (e.g.: Uber: https://eng.uber.com/go-monorepo-bazel/)

File: BUILD.bazel [hunk 1]
```diff
@@ -8,6 +8,16 @@ gazelle(
     prefix = "github.com/buchgr/bazel-remote",
 )
 
+gazelle(
+    name = "gazelle-update-repos",
+    args = [
+        "-from_file=go.mod",
+        "-to_macro=deps.bzl%go_dependencies",
+        "-prune",
```
Comment (mostynb) [inline BUILD.bazel:16]: This looks like it removed a few entries already, nice.

**Review by mostynb — APPROVED** [review]
OK, merging- thanks.

**Pull request merged**